### ISSUE TITLE
Compilation fails for models with non-long mask types from PyTorch

### ISSUE THREAD

**Issue by MaxS1996** [iree-org/iree#20549]
### What happened?

I encountered an issue where the compilation of a model that utilizes masking fails if the mask has any type other than torch.long. PyTorch supports bool, int, and long (int64) for the mask, but only long can be lowered successfully when compiling these models with IREE.

```python
import torch
import torch.nn as nn

class MaskingModule(nn.Module):
    
    def __init__(self, dtype_data=torch.float, dtype_mask=torch.long):
        super(MaskingModule, self).__init__()
        self.dtype_data = dtype_data
        self.dtype_mask = dtype_mask

    def forward(self, x):
        tensor_a = torch.zeros_like(x, dtype=self.dtype_data)
        tensor_b = torch.ones_like(x, dtype=self.dtype_data) * 2

        mask = torch.rand_like(x) * 10
        mask = mask < 5
        mask = mask.to(self.dtype_mask)

        tensor_a[mask] = tensor_b[mask]
        return tensor_a

# Example usage
shape = (2, 100, 3)
dtype_data = torch.float32
dtype_mask = torch.bool
x = torch.rand(shape)
example_args = (x,)
model_torch = MaskingModule(dtype_data, dtype_mask)
out = model_torch(*example_args)

import iree.turbine.aot as aot

export_output = aot.export(model_torch, *example_args)
binary = export_output.compile(
    target_backends=["llvm-cpu"], 
    save_to="test.vmfb"
)

```

I assume that there is some issue with the type associations of the tensor indicies, which always seem to be i64 in torch-mlir and s(?)i32 in linalg.
An explicit conversion to torch.long in PyTorch will result in redunant ops throughout the lowered stages, where the mask is firstly converted to i64 and than truncated to i32 in two separate linalg.generic steps

### Steps to reproduce your issue

1. Define a model that uses masking with a mask type other than torch.long.
i.e. MaskingModule with dtype_data = torch.float32 and dtype_mask = torch.bool
2. Compile it to linalg using iree-turbine works for all configurations
3. Compiling the model to a VM flatbuffer using the compile step fails

Observed Behavior:

* Mask dtype = torch.long/torch.int64: Compiles and executes without issues.
* Mask dtype = torch.int/torch.int32: Compilation fails with the error: 'arith.cmpi' op requires all operands to have the same type1.
* Mask dtype = torch.bool: Compilation does not fail, but it never terminates and keeps allocating memory until the system runs out (tested on a machine with more than 128GB

### What component(s) does this issue relate to?

Compiler

### Version information

IREE 3.3.0 for compiler, iree-turbine, iree-runtime (not used)

### Additional context

I am currently looking into architectures that operate on (LIDAR) point clouds, like PointNet++. These kind of masked assign operations are quite common in this problem class.

**Comment by MaxS1996** [comment]
might be related to #20550
